File: 1036-rotting-oranges/rotting-oranges.java
class Solution {
    public int orangesRotting(int[][] grid) {

        int rowLength = grid.length, colsLength = grid[0].length;
        int time = 0;
        Queue<int[]> q = new LinkedList<>();

        for (int i = 0; i < rowLength; i++) {
            for (int j = 0; j < colsLength; j++) {
                if (grid[i][j] == 2) {
                    q.add(new int[] { i, j });
                }
            }
        }

        int[][] directions = { { 0, 1 }, { 1, 0 }, { -1, 0 }, { 0, -1 } };
        while (!q.isEmpty()) {
            boolean rotted = false;
            int qSize = q.size();
            for (int i = 0; i < qSize; i++) {
                int[] curr = q.poll();
                int row = curr[0];
                int col = curr[1];
                for (int[] direction : directions) {
                    int newRow = row + direction[0];
                    int newCol = col + direction[1];
                    if (newRow >= 0 && newCol >= 0 && newRow < rowLength && newCol < colsLength &&
                            grid[newRow][newCol] == 1) {
                        grid[newRow][newCol] = 2;
                        q.add(new int[] { newRow, newCol });
                        rotted = true;
                    }
                }
            }
            if (rotted) {
                time++;
            }
        }

        for (int i = 0; i < rowLength; i++) {
            for (int j = 0; j < colsLength; j++) {
                if (grid[i][j] == 1) {
                    return -1;
                }
            }
        }
        return time;
    }
}

File: 1036-rotting-oranges/rotting-oranges.cpp
class Solution {
public:
    int orangesRotting(vector<vector<int>>& grid) {
        int n = grid.size();  // Get the number of rows
        int m = grid[0].size();  // Get the number of columns
        
        // Queue to store the rotten oranges and their time stamp { {row, col}, time }
        queue<pair<pair<int, int>, int>> q;

        // Visited matrix to track which oranges are already processed
        vector<vector<int>> vis(n, vector<int>(m, 0));

        // Step 1: Push all initially rotten oranges into the queue
        for(int i = 0; i < n; i++) {
            for(int j = 0; j < m; j++) {
                if(grid[i][j] == 2) {  // Rotten orange found
                    q.push({{i, j}, 0}); // Store position and time (0 initially)
                    vis[i][j] = 2;  // Mark it as visited using 2
                }
            }
        }

        int time = 0;  // Track the time required to rot all oranges

        // Direction arrays for moving up, right, down, left
        int drow[] = {-1, 0, 1, 0};
        int dcol[] = {0, 1, 0, -1};

        // Step 2: Perform BFS traversal to rot fresh oranges
        while(!q.empty()) {
            int r = q.front().first.first;  // Current row
            int c = q.front().first.second; // Current column
            int t = q.front().second;  // Current time
            q.pop();

            time = max(time, t);  // Update the time taken

            // Traverse all 4 possible directions
            for(int i = 0; i < 4; i++) {
                int nrow = r + drow[i];  // New row
                int ncol = c + dcol[i];  // New column

                // Check if the new position is within bounds and contains a fresh orange
                if(nrow >= 0 && nrow < n && ncol >= 0 && ncol < m && vis[nrow][ncol] == 0 && grid[nrow][ncol] == 1) {
                    q.push({{nrow, ncol}, t + 1});  // Push newly rotten orange with updated time
                    vis[nrow][ncol] = 2;  // Mark it as visited (rotten)
                    grid[nrow][ncol] = 2; // Update grid to indicate it's now rotten
                }
            }
        }

        // Step 3: Check if any fresh oranges are left
        for(int i = 0; i < n; i++) {
            for(int j = 0; j < m; j++) {
                if(grid[i][j] == 1) {  // If any fresh orange is left, return -1
                    return -1;
                }
            }
        }

        return time;  // Return the total time taken to rot all oranges
    }
};